public/app/features/alerting/unified/utils/notification-policies.ts

```diff
@@ -11,111 +11,84 @@ import { Labels } from 'app/types/unified-alerting-dto';
 
 import { Label, normalizeMatchers } from './matchers';
 
-type OperatorPredicate = (labelValue: string, matcherValue: string) => boolean;
-
-const OperatorFunctions: Record<MatcherOperator, OperatorPredicate> = {
-  [MatcherOperator.equal]: (lv, mv) => lv === mv,
-  [MatcherOperator.notEqual]: (lv, mv) => lv !== mv,
-  [MatcherOperator.regex]: (lv, mv) => Boolean(lv.match(new RegExp(mv))),
-  [MatcherOperator.notRegex]: (lv, mv) => !Boolean(lv.match(new RegExp(mv))),
-};
-
-function isLabelMatch(matcher: ObjectMatcher, label: Label) {
-  const [labelKey, labelValue] = label;
-  const [matcherKey, operator, matcherValue] = matcher;
-
-  // not interested, keys don't match
-  if (labelKey !== matcherKey) {
-    return false;
-  }
-
-  const matchFunction = OperatorFunctions[operator];
-  if (!matchFunction) {
-    throw new Error(`no such operator: ${operator}`);
-  }
-
-  return matchFunction(labelValue, matcherValue);
-}
-
+// If a policy has no matchers it still can be a match, hence matchers can be empty and match can be true
+// So we cannot use null as an indicator of no match
 interface LabelMatchResult {
   match: boolean;
-  matchers: ObjectMatcher[];
+  matcher: ObjectMatcher | null;
```

Comment:
In the preview routing, we employ this type to indicate which matchers are associated with each label in a matching instance within a route. From a technical perspective, it's indeed possible. For instance, in the following scenario, this label matches due to multiple matchers:

In this case, the `alertname=rule5` label matches 2 different matchers:

<img width="1173" alt="Screenshot 2023-10-30 at 12 44 41" src="https://github.com/grafana/grafana/assets/33540275/a3adc9cc-2bdc-44c1-ba9a-4c59b14afcc2">

Reply:
That makes sense, from an algorithm's perspective there's only ever one matcher that decides for a label. It's either the first matcher that it doesn't match or the last matcher that does. In other words, for the following set of matchers and alert instance with `alertname=foo`

`alertname=~.*, alertname=foo` it will be the last matcher (`alertname=foo`)

`alertname!=bar, alertname=foo` the matcher will be `alertname!=foo` since it will halt evaluating other matchers after a negative match.

It looks like we're only visualizing the label that matched and not the matchers for that label so I don't think it matters all that much here anyway :)

Reply:
Thanks Gilles!
yes, you're right. I was reviewing the code and it seems we are not using it, so, this change is not going to add a regression 👍